src/Streamly/Coreutils/Cp.hs

```diff
@@ -0,0 +1,258 @@
+       asFile <- parseRelFile path
+       let filePath = dir </> asFile
+       isFile <- doesFileExist $ fromAbsFile filePath
+       if isFile == True
+       then do
+           return $ Right filePath
+       else do
+           asDir <- parseRelDir path
+           return $ Left $ dir </> asDir
+
+    contents :: FilePath -> [FilePath]
+    contents dir = unsafePerformIO $ listDirectory dir
+
+
+
+-- | Recursively copies the contents of the source directory
+-- to the destination directory.
+--
+-- @since 0.1.0.0
+copyDirToDir
+    :: IsStream t
+    => CpOptions
+    -> Path Abs Dir
+    -> Path Abs Dir
+    -> t IO ()
+copyDirToDir opt src dest = do
+    if force opt == True
+    then
+        S.mapM handleForce
+        $ traverseDir src
+    else
+        S.mapM (\_ -> return ())
+        $ S.takeWhileM handleNonForce
+        $ traverseDir src
+
+    where
+
+    handleForce :: Either (Path Abs Dir) (Path Abs File) -> IO ()
+    handleForce (Left dir) = do
+        if dir == src
+        then createDirectoryIfMissing True $ fromAbsDir dest
+        else do
+            actualPath <- stripAndPrepend dir src dest
+            createDirectoryIfMissing True $ fromAbsDir actualPath
+    handleForce (Right file) = do
+        actualPath <- stripAndPrepend file src dest
+        _ <- copy opt file actualPath
+        return ()
+
+
+    handleNonForce :: Either (Path Abs Dir) (Path Abs File) -> IO Bool
+    handleNonForce (Left dir) = do
+        if dir == src
+        then do
+            createDirectoryIfMissing True $ fromAbsDir dest
+            return True
+        else do
+            actualPath <- stripAndPrepend dir src dest
+            createDirectoryIfMissing True $ fromAbsDir actualPath
+            return True
+    handleNonForce (Right file) = do
+        actualPath <- stripAndPrepend file src dest
+        copy opt file actualPath
+
+    stripAndPrepend :: Path Abs t -> Path Abs Dir -> Path Abs Dir  -> IO (Path Abs t)
+    stripAndPrepend path strip pref = do
+        suffix <- stripProperPrefix strip path
+        return $ pref </> suffix
+
+
+
+
+-- | Copies the source file preserving the source file name
+-- in the destination directory.
+-- Returns @True@ in the @IO@ monad if copying was successful,
+-- @False@ otherwise.
+--
+-- @since 0.1.0.0
+cpFile
+    :: CpOptions
+    -> Path Abs File
+    -> Path Abs Dir
+    -> IO Bool
+cpFile option src dest =
+    copy option src $ dest </> filename src
+
+
+-- | Copies a stream of files to the destination directory.
+-- The stream can be drained with the appropriate combinator
+-- to achieve concurrency.
+-- Fails if the destination directory does not exist.
+--
+-- @since 0.1.0.0
+cpFiles
+    :: IsStream t
+    => CpOptions
+    -> t IO (Path Abs File)
+    -> Path Abs Dir
+    -> t IO (Path Abs File)
+    -- ^ Stream of files for which copying was successful
+cpFiles opt strm dir = do
+    if force opt == True
+    then
+        S.filterM (\f -> cpFile opt f dir) strm
```

Comment:
Why is one of the case using `filterM` and the other `takeWhileM` and how is this related to `force` option.